pydm/data_plugins/calc_plugin.py

```diff
@@ -0,0 +1,204 @@
+        QThread.__init__(self, *args, **kwargs)
+        self.app = QApplication.instance()
+        self.app.aboutToQuit.connect(self.requestInterruption)
+
+        self.config = config
+
+        self._calculate = threading.Event()
+        self._channels = []
+        self._value = None
+        self._values = None
+        self._connections = None
+        self._expression = self.config.get('expr', '')
+
+        for ch_idx, ch in enumerate(self.config.get('channels', [])):
+            conn_cb = functools.partial(self.callback_conn, ch_idx)
+            value_cb = functools.partial(self.callback_value, ch_idx)
+            c = pydm.PyDMChannel(ch, connection_slot=conn_cb,
+                                 value_slot=value_cb)
+            self._channels.append(c)
+
+    @property
+    def connected(self):
+        return all(self._connections)
+
+    def _connect(self):
+        self._values = [None]*len(self._channels)
+        self._connections = [False]*len(self._channels)
+        for ch in self._channels:
+            ch.connect()
+
+    def _disconnect(self):
+        for ch in self._channels:
+            ch.disconnect()
+
+    def _send_update(self, conn, value):
+        self.new_data_signal.emit({"connection": conn,
+                                   "value": value})
+
+    def run(self):
+        self._connect()
+
+        while True:
+            self._calculate.wait()
+            self._calculate.clear()
+            if self.isInterruptionRequested():
+                break
+            self.calculate_expression()
+        self._disconnect()
+
+    def callback_value(self, ch_index, value):
+        """
+        Callback executed when a channel receives a new value.
+
+        Parameters
+        ----------
+        ch_index : int
+            The channel index on the list for this rule.
+        value : any
+            The new value for this channel.
+
+        Returns
+        -------
+        None
+        """
+        self._values[ch_index] = value
+        if not self.connected:
+            self.warn_unconnected_channels()
+            return
+        self._calculate.set()
+
+
+    def callback_conn(self, ch_index, value):
+        """
+        Callback executed when a channel connection status is changed.
+
+        Parameters
+        ----------
+        ch_index : int
+            The channel index on the list for this rule.
+        value : bool
+            Whether or not this channel is connected.
+
+        """
+        self._connections[ch_index] = value
+        self._send_update(self.connected, self._value)
+
+    def warn_unconnected_channels(self):
+        logger.debug(
+            "Calculation '%s': Not all channels are connected, skipping execution.",
+            self.objectName())
+
+    def calculate_expression(self):
+        """
+        Evaluate the expression defined by the rule and emit the `rule_signal`
+        with the new value.
+        """
+        vals = list(self._values)
```

Comment:
`calc://{"name":"circley", "channels":[ca://DEMO:ANGLE], "expr":"math.sin(math.radians(ch[0]))"}`

Could be made a bit less awkward with:
`calc://{"name":"circley", "channels":{"angle":"ca://DEMO:ANGLE"}, "expr":"math.sin(math.radians(angle))"}`

That is, letting the user decide the variable names.
Then below, you would be doing `env.update(**user_variable_value_dict)` or something along those lines.

Reply:
This seems like a great idea! It was originally using the index as I was replicating the same feel as from the rules.

Reply:
Addressed via https://github.com/slaclab/pydm/pull/637/commits/c9daee59422ea74d03dfa8ff3c76968a0a6492fa. It turned out way better and leveraging also `defaultdict` allowed me to remove some other lines of code.

Reply:
Awesome 👍